hoomd/data/data_structures.py

```diff
@@ -0,0 +1,497 @@
+from contextlib import contextmanager
+from copy import deepcopy
+from itertools import cycle
+
+from hoomd.data.typeconverter import (
+    TypeConversionError, TypeConverterValue, OnlyIf, Either,
+    TypeConverterMapping, TypeConverterSequence)
+
+
+class _HOOMDDataStructures(metaclass=ABCMeta):
+    """Base class for HOOMD's base container classes.
+
+    These child classes exist to enable parameter validation and mantain
+    consistency with C++ after attaching.
+    """
+    @abstractmethod
+    def to_base(self):
+        """Cast the object into the corresponding native Python container type.
+        """
+        pass
+
+    @contextmanager
+    def _buffer(self):
+        """Allows buffering during modifications to prevent needless updates.
+
+        This is a simple system to prevent unnecessary C++ - Python
+        communication.
+        """
+        self._buffered = True
+        yield None
+        self._buffered = False
+
+    def _update(self):
+        """Signal that the object has been updated."""
+        if self._buffered:
+            return
+        elif parent is not None:
+            self._parent._handle_update(self, self._label)
+
+
+def _get_inner_typeconverter(type_def, desired_type):
+    """Handles more complicated data structure specification.
+
+    When using nested containers where an entry could be a list or None for
+    instance requires the use of something like typeconverter's OnlyIf or
+    Either. We autdetect the internal type specification for these object and if
+    they are of the correct type use these.
+
+    Args:
+        type_def (`hoomd.data.typeconverter.TypeConverter`):
+            The type converter which is desired to be introspected to find the
+            desired type.
+        desired_type (`hoomd.data.typeconverter.TypeConverter`):
+            A child class of `hoomd.data.typeconverter.TypeConverter` that
+            signals what type converter types is needed.
+    """
+    rtn_type_def = None
+    # Simple case the type_def is already sufficient
+    if isinstance(type_def, desired_type):
+        rtn_type_def = type_def
+    # More complicated cases, deals with nested type definitions where the
+    # nested containers have other requirements such as multiple valid container
+    # representations or entry is allowed to be None. In these cases, we check
+    # the TypeConverterValue instance's converter for the desired container type
+    # converter type.
+    elif isinstance(type_def, TypeConverterValue):
+        if isinstance(type_def.converter, OnlyIf):
+            if isinstance(type_def.converter.cond, desired_type):
+                rtn_type_def = type_def.converter.cond
+        elif isinstance(type_def.converter, Either):
+            for spec in type_def.converter.specs:
+                matches = []
+                if isinstance(spec, desired_type):
+                    matches.append(spec)
+                if len(matches) > 1:
```

Comment:
This can't happen, check the logic -- `matches = []` happens inside the for loop. I attempt to resolve this in my rewrite below.